### PR TITLE
Support passing a relative string to pushState/replaceState

### PR DESCRIPTION
**Pull request by timneutkens**
Follow-up to #58335.

Fixes https://github.com/vercel/next.js/discussions/48110#discussioncomment-7565394

As reported in the discussion passing a origin-relative string didn't work as `new URL` will throw an error. This ensures the `origin` parameter is provided to `new URL`.
Added tests for the string behavior.

### CODE REVIEW

**Comment by timneutkens** [comment]
Current dependencies on/for this PR:
* canary
  * **PR #58426** <a href="https://app.graphite.dev/github/pr/vercel/next.js/58426?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #58438** <a href="https://app.graphite.dev/github/pr/vercel/next.js/58438?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/vercel/next.js/58438?utm_source=stack-comment).

File: packages/next/src/client/components/app-router.tsx [hunk 1]
```diff
@@ -207,13 +207,13 @@ function useChangeByServerResponse(
 function useNavigate(dispatch: React.Dispatch<ReducerActions>): RouterNavigate {
   return useCallback(
     (href, navigateType, forceOptimisticNavigation, shouldScroll) => {
-      const url = new URL(addBasePath(href), location.href)
+      const url = new URL(addBasePath(href), window.location.href)
```
Comment (timneutkens) [inline packages/next/src/client/components/app-router.tsx:210]: For consistency with other cases.

Comment (franky47) [reply packages/next/src/client/components/app-router.tsx:210]: Could `location.*` (and conversely, `history.*` too) be used everywhere? Surely repeating window increases the bundle size (albeit it probably compresses better).

**Review by ztanner — DISMISSED** [review]

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js 11-14-Support_passing_a_relative_string_to_pushState/replaceState | Change |
| - | - | - | - |
| buildDuration | 10.8s | 10.7s | N/A |
| buildDurationCached | 6s | 6.1s | N/A |
| nodeModulesSize | 199 MB | 199 MB | ⚠️ +1.19 kB |
| nextStartRea..uration (ms) | 415ms | 415ms | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-14-Support_passing_a_relative_string_to_pushState/replaceState | Change |
| - | - | - | - |
| 199-HASH.js gzip | 29.2 kB | 29.3 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | N/A |
| 494.HASH.js gzip | 180 B | 181 B | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 243 B | 239 B | N/A |
| main-HASH.js gzip | 31.7 kB | 31.8 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 46.9 kB | 46.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-14-Support_passing_a_relative_string_to_pushState/replaceState | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-14-Support_passing_a_relative_string_to_pushState/replaceState | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 504 B | 506 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | ✓ |
| edge-ssr-HASH.js gzip | 253 B | 255 B | N/A |
| head-HASH.js gzip | 348 B | 347 B | N/A |
| hooks-HASH.js gzip | 369 B | 368 B | N/A |
| image-HASH.js gzip | 4.3 kB | 4.3 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.65 kB | 2.65 kB | N/A |
| routerDirect..HASH.js gzip | 311 B | 311 B | ✓ |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.17 kB | 3.17 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-14-Support_passing_a_relative_string_to_pushState/replaceState | Change |
| - | - | - | - |
| _buildManifest.js gzip | 486 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-14-Support_passing_a_relative_string_to_pushState/replaceState | Change |
| - | - | - | - |
| index.html gzip | 529 B | 527 B | N/A |
| link.html gzip | 541 B | 543 B | N/A |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-14-Support_passing_a_relative_string_to_pushState/replaceState | Change |
| - | - | - | - |
| edge-ssr.js gzip | 92.6 kB | 92.6 kB | N/A |
| page.js gzip | 145 kB | 145 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-14-Support_passing_a_relative_string_to_pushState/replaceState | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 626 B | N/A |
| middleware-r..fest.js gzip | 150 B | 151 B | N/A |
| middleware.js gzip | 24.8 kB | 24.8 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 1.92 kB | 1.92 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-14-Support_passing_a_relative_string_to_pushState/replaceState | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 167 kB | 167 kB | ✓ |
| app-page-exp..prod.js gzip | 93.2 kB | 93.2 kB | ✓ |
| app-page-tur..prod.js gzip | 93.9 kB | 93.9 kB | ✓ |
| app-page-tur..prod.js gzip | 88.5 kB | 88.5 kB | ✓ |
| app-page.run...dev.js gzip | 137 kB | 137 kB | ✓ |
| app-page.run..prod.js gzip | 87.9 kB | 87.9 kB | ✓ |
| app-route-ex...dev.js gzip | 23.8 kB | 23.8 kB | ✓ |
| app-route-ex..prod.js gzip | 16.4 kB | 16.4 kB | ✓ |
| app-route-tu..prod.js gzip | 16.4 kB | 16.4 kB | ✓ |
| app-route-tu..prod.js gzip | 16 kB | 16 kB | ✓ |
| app-route.ru...dev.js gzip | 23.2 kB | 23.2 kB | ✓ |
| app-route.ru..prod.js gzip | 16 kB | 16 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.64 kB | 9.64 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-turbo...prod.js gzip | 21.8 kB | 21.8 kB | ✓ |
| pages.runtim...dev.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim..prod.js gzip | 21.8 kB | 21.8 kB | ✓ |
| server.runti..prod.js gzip | 49 kB | 49 kB | ✓ |
| Overall change | 923 kB | 923 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>199-HASH.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 585790ab4bcc14f86821982ae94dc277942b9db3

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 585790ab4bcc14f86821982ae94dc277942b9db3

`pnpm test-dev test/development/middleware-errors/index.test.ts` 
- middleware - development errors > when there is a compilation error from boot > logs the error correctly

<details>
<summary>Expand output</summary>

  ● middleware - development errors › when there is a compilation error from boot › logs the error correctly

    TIMED OUT: success

    undefined

    Error: expect(received).toEqual(expected) // deep equality

    Expected: 2
    Received: 3

      626 |
      627 |   if (hardError) {
    > 628 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      629 |   }
      630 |   return false
      631 | }

      at check (lib/next-test-utils.ts:628:11)
      at Object.<anonymous> (development/middleware-errors/index.test.ts:250:9)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by feedthejim — DISMISSED** [review]

File: packages/next/src/client/components/app-router.tsx [hunk 2]
```diff
@@ -404,12 +404,14 @@ function Router({
...
-      )
+      ) {
         return
+      }
 
+      const href = window.location.href
       dispatch({
         type: ACTION_RESTORE,
-        url: new URL(window.location.href),
+        url: new URL(href, href),
```
Comment (dvoytenko) [inline packages/next/src/client/components/app-router.tsx:414]: No need for `href, href` here.

File: packages/next/src/client/components/app-router.tsx [hunk 3]
```diff
@@ -525,9 +528,10 @@ function Router({
       // TODO-APP: Ideally the back button should not use startTransition as it should apply the updates synchronously
       // Without startTransition works if the cache is there for this path
       startTransition(() => {
+        const href = window.location.href
         dispatch({
           type: ACTION_RESTORE,
-          url: new URL(window.location.href),
+          url: new URL(href, href),
```
Comment (dvoytenko) [inline packages/next/src/client/components/app-router.tsx:534]: No need for `href, href`

File: test/e2e/app-dir/shallow-routing/app/(shallow)/pushstate-string-url/page.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  return (
+    <>
+      <h1 id="pushstate-string-url">PushState String Url</h1>
+      <pre id="my-data">{searchParams.get('query')}</pre>
+      <button
+        onClick={() => {
+          const previousQuery = new URL(window.location.href).searchParams.get(
+            'query'
+          )
+          const url = `${window.location.pathname}?query=${
```
Comment (dvoytenko) [inline test/e2e/app-dir/shallow-routing/app/(shallow)/pushstate-string-url/page.tsx:15]: Just do `?query=...`. It'd be more clear for this test.

**Review by dvoytenko — CHANGES_REQUESTED** [review]

**Comment by dvoytenko** [comment]
@timneutkens while we're here, can we fix the https://github.com/vercel/next.js/pull/58335#discussion_r1396095644 as well?

**Review by dvoytenko — APPROVED** [review]

**Review by ztanner — DISMISSED** [review]

File: packages/next/src/client/components/app-router.tsx [hunk 5]
```diff
@@ -232,6 +232,7 @@ const originalReplaceState =
     : null
 
 function copyNextJsInternalHistoryState(data: any) {
+  if (data == null || data === undefined) data = {}
```
Comment (timneutkens) [inline packages/next/src/client/components/app-router.tsx:235]: `data == null` covers `undefined`:

```
(undefined == null) === true
```

**Review by ztanner — APPROVED** [review]

**Comment by cduff** [comment]
FYI, looks like this will resolve https://github.com/vercel/next.js/issues/56636.

**Pull request merged**